src/pool/mempool.rs

```diff
@@ -21,95 +33,146 @@ pub type KakarotPool<Client> = Pool<Validator<Client>, TransactionOrdering, Noop
+    accounts: Arc<Mutex<HashMap<Felt, Felt>>>,
 }
 
 impl AccountManager {
-    pub fn new(path: &str) -> Self {
-        let mut accounts = HashSet::new();
+    pub async fn new<SP>(path: &str, eth_provider: &EthDataProvider<SP>) -> eyre::Result<Self>
+    where
+        SP: starknet::providers::Provider + Send + Sync + 'static,
+    {
+        let mut accounts = HashMap::new();
 
         // Open the file specified by `path`
         let Ok(mut file) = File::open(path) else {
-            return Self::default();
+            return Ok(Self::default());
         };
 
         let mut contents = String::new();
         if file.read_to_string(&mut contents).is_err() {
-            return Self::default();
+            return Ok(Self::default());
         }
 
         // Parse the file contents as JSON
         let json: Value = match serde_json::from_str(&contents) {
             Ok(json) => json,
             Err(_) => {
-                return Self::default();
+                return Ok(Self::default());
             }
         };
 
         // Extract the account addresses from the JSON array
         if let Some(array) = json.as_array() {
             for item in array {
-                if let Some(address) = item.as_str() {
-                    accounts.insert(Felt::from_hex_unchecked(address));
+                if let Some(account_address) = item.as_str() {
+                    let felt_address = Felt::from_hex_unchecked(account_address);
+
+                    let starknet_block_id = eth_provider
+                        .to_starknet_block_id(Some(BlockId::default()))
+                        .await
+                        .map_err(|e| eyre::eyre!("Error converting block ID: {:?}", e))?;
+
+                    // Query the initial account_nonce for the account from the provider
+                    accounts.insert(
+                        felt_address,
+                        eth_provider
+                            .starknet_provider()
+                            .get_nonce(starknet_block_id, felt_address)
+                            .await
+                            .unwrap_or_default(),
+                    );
                 }
             }
         }
 
-        Self { accounts }
+        Ok(Self { accounts: Arc::new(Mutex::new(accounts)) })
     }
 
-    pub fn start<SP>(&self, rt_handle: &Handle, eth_client: &'static EthClient<SP>)
+    /// Start the account manager task.
+    #[allow(clippy::significant_drop_tightening)]
+    pub fn start<SP>(&self, rt_handle: &Handle, eth_client: Arc<EthClient<SP>>)
     where
         SP: starknet::providers::Provider + Send + Sync + Clone + 'static,
     {
         let accounts = self.accounts.clone();
 
         rt_handle.spawn(async move {
             loop {
-                for address in &accounts {
-                    Self::process_transaction(address, eth_client);
+                let result: eyre::Result<()> = {
+                    let mut accounts = accounts.lock().await;
```

Comment:
should we already take lock here? I think we can iterate the accounts as a view, check if balance is ok and then take a lock only when needed no?

Reply:
You are right, just modify. We can go without acquiring the lock at the beginning, during balance check and so on. And then only acquiring the lock when we need to modify the hashmap within the `process_transaction` method.